File: go-tsdb/internal/storage/storage_test.go
package storage

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"github.com/vobbilis/go-tsdb/test/testutil"
)

func TestStorageBasicOperations(t *testing.T) {
	storage, err := NewStorage(StorageConfig{
		DataPath:     t.TempDir(),
		MaxBlockSize: 1024 * 1024,
	})
	require.NoError(t, err)
	defer storage.Close()

	// Test data
	series := testutil.GenerateTestSeries(map[string]string{
		"__name__": "test_metric",
		"instance": "test-1",
	}, 100)

	// Test Write
	t.Run("Write", func(t *testing.T) {
		err := storage.WriteSeries(series)
		assert.NoError(t, err)
	})

	// Test Read
	t.Run("Read", func(t *testing.T) {
		end := time.Now()
		start := end.Add(-1 * time.Hour)

		result, err := storage.ReadSeries(series.Labels, start.UnixNano(), end.UnixNano())
		assert.NoError(t, err)
		assert.NotNil(t, result)
		assert.Equal(t, len(series.Samples), len(result.Samples))
	})

	// Test Query
	t.Run("Query", func(t *testing.T) {
		matcher := LabelMatcher{
			Type:  MatchEqual,
			Name:  "__name__",
			Value: "test_metric",
		}

		end := time.Now()
		start := end.Add(-1 * time.Hour)

		results, err := storage.Query([]LabelMatcher{matcher}, start.UnixNano(), end.UnixNano())
		assert.NoError(t, err)
		assert.NotEmpty(t, results)
	})
}

func TestStorageConcurrency(t *testing.T) {
	storage, err := NewStorage(StorageConfig{
		DataPath:     t.TempDir(),
		MaxBlockSize: 1024 * 1024,
	})
	require.NoError(t, err)
	defer storage.Close()

	// Generate test data
	testData := testutil.GenerateTestData(10, 1000)

	// Test concurrent writes
	t.Run("ConcurrentWrites", func(t *testing.T) {
		done := make(chan error, len(testData))

		for _, series := range testData {
			go func(s *types.Series) {
				done <- storage.WriteSeries(s)
			}(series)
		}

		// Collect results
		for i := 0; i < len(testData); i++ {
			err := <-done
			assert.NoError(t, err)
		}
	})

	// Test concurrent reads
	t.Run("ConcurrentReads", func(t *testing.T) {
		end := time.Now()
		start := end.Add(-1 * time.Hour)

		done := make(chan error, len(testData))

		for _, series := range testData {
			go func(s *types.Series) {
				_, err := storage.ReadSeries(s.Labels, start.UnixNano(), end.UnixNano())
				done <- err
			}(series)
		}

		// Collect results
		for i := 0; i < len(testData); i++ {
			err := <-done
			assert.NoError(t, err)
		}
	})
}

func TestStorageCompaction(t *testing.T) {
	storage, err := NewStorage(StorageConfig{
		DataPath:     t.TempDir(),
		MaxBlockSize: 1024, // Small block size to trigger compaction
	})
	require.NoError(t, err)
	defer storage.Close()

	// Write enough data to trigger compaction
	series := testutil.GenerateTestSeries(map[string]string{
		"__name__": "test_metric",
		"instance": "test-1",
	}, 10000)

	t.Run("WriteAndCompact", func(t *testing.T) {
		err := storage.WriteSeries(series)
		assert.NoError(t, err)

		// Wait for compaction
		time.Sleep(2 * time.Second)

		// Verify data is still readable
		end := time.Now()
		start := end.Add(-1 * time.Hour)

		result, err := storage.ReadSeries(series.Labels, start.UnixNano(), end.UnixNano())
		assert.NoError(t, err)
		assert.NotNil(t, result)
		assert.Equal(t, len(series.Samples), len(result.Samples))
	})
}

func TestStorageRecovery(t *testing.T) {
	dir := t.TempDir()

	// Create and populate storage
	storage, err := NewStorage(StorageConfig{
		DataPath:     dir,
		MaxBlockSize: 1024 * 1024,
	})
	require.NoError(t, err)

	series := testutil.GenerateTestSeries(map[string]string{
		"__name__": "test_metric",
		"instance": "test-1",
	}, 100)

	err = storage.WriteSeries(series)
	require.NoError(t, err)

	// Close storage
	storage.Close()

	// Reopen storage
	storage2, err := NewStorage(StorageConfig{
		DataPath:     dir,
		MaxBlockSize: 1024 * 1024,
	})
	require.NoError(t, err)
	defer storage2.Close()

	// Verify data survived
	t.Run("RecoveredData", func(t *testing.T) {
		end := time.Now()
		start := end.Add(-1 * time.Hour)

		result, err := storage2.ReadSeries(series.Labels, start.UnixNano(), end.UnixNano())
		assert.NoError(t, err)
		assert.NotNil(t, result)
		assert.Equal(t, len(series.Samples), len(result.Samples))
	})
}

File: cpp-tsdb/test/storage/storage_test.cpp
#include <gtest/gtest.h>
#include <filesystem>
#include <thread>
#include <vector>
#include "tsdb/storage/storage.hpp"
#include "../test_utils.hpp"

namespace tsdb::test {

class StorageTest : public ::testing::Test {
protected:
    void SetUp() override {
        // Create temporary directory for tests
        test_path_ = std::filesystem::temp_directory_path() / "tsdb_test";
        std::filesystem::create_directories(test_path_);
        
        // Initialize storage
        StorageConfig config{
            .data_path = test_path_.string(),
            .max_block_size = 1024 * 1024
        };
        storage_ = std::make_unique<Storage>(config);
    }

    void TearDown() override {
        storage_.reset();
        std::filesystem::remove_all(test_path_);
    }

    std::filesystem::path test_path_;
    std::unique_ptr<Storage> storage_;
};

TEST_F(StorageTest, BasicOperations) {
    // Test data
    auto series = TestUtils::generateTestSeries({
        {"__name__", "test_metric"},
        {"instance", "test-1"}
    }, 100);

    // Test Write
    EXPECT_NO_THROW(storage_->writeSeries(series));

    // Test Read
    auto now = std::chrono::system_clock::now();
    auto start = now - std::chrono::hours(1);
    auto result = storage_->readSeries(
        series.getLabels(),
        std::chrono::duration_cast<std::chrono::nanoseconds>(start.time_since_epoch()).count(),
        std::chrono::duration_cast<std::chrono::nanoseconds>(now.time_since_epoch()).count()
    );

    ASSERT_TRUE(result);
    EXPECT_EQ(result->getSamples().size(), series.getSamples().size());

    // Test Query
    LabelMatcher matcher{
        .type = MatchType::Equal,
        .name = "__name__",
        .value = "test_metric"
    };

    auto results = storage_->query(
        {matcher},
        std::chrono::duration_cast<std::chrono::nanoseconds>(start.time_since_epoch()).count(),
        std::chrono::duration_cast<std::chrono::nanoseconds>(now.time_since_epoch()).count()
    );

    EXPECT_FALSE(results.empty());
}

TEST_F(StorageTest, Concurrency) {
    // Generate test data
    auto testData = TestUtils::generateTestData(10, 1000);
    
    // Test concurrent writes
    {
        std::vector<std::thread> threads;
        std::atomic<int> errors{0};
        
        for (const auto& series : testData) {
            threads.emplace_back([this, &series, &errors]() {
                try {
                    storage_->writeSeries(series);
                } catch (...) {
                    errors++;
                }
            });
        }

        for (auto& thread : threads) {
            thread.join();
        }

        EXPECT_EQ(errors, 0);
    }

    // Test concurrent reads
    {
        auto now = std::chrono::system_clock::now();
        auto start = now - std::chrono::hours(1);
        auto startNs = std::chrono::duration_cast<std::chrono::nanoseconds>(
            start.time_since_epoch()
        ).count();
        auto endNs = std::chrono::duration_cast<std::chrono::nanoseconds>(
            now.time_since_epoch()
        ).count();

        std::vector<std::thread> threads;
        std::atomic<int> errors{0};
        
        for (const auto& series : testData) {
            threads.emplace_back([this, &series, startNs, endNs, &errors]() {
                try {
                    auto result = storage_->readSeries(series.getLabels(), startNs, endNs);
                    if (!result) {
                        errors++;
                    }
                } catch (...) {
                    errors++;
                }
            });
        }

        for (auto& thread : threads) {
            thread.join();
        }

        EXPECT_EQ(errors, 0);
    }
}

TEST_F(StorageTest, Compaction) {
    // Configure storage with small block size to trigger compaction
    StorageConfig config{
        .data_path = test_path_.string(),
        .max_block_size = 1024  // Small block size
    };
    storage_ = std::make_unique<Storage>(config);

    // Write enough data to trigger compaction
    auto series = TestUtils::generateTestSeries({
        {"__name__", "test_metric"},
        {"instance", "test-1"}
    }, 10000);

    EXPECT_NO_THROW(storage_->writeSeries(series));

    // Wait for compaction
    std::this_thread::sleep_for(std::chrono::seconds(2));

    // Verify data is still readable
    auto now = std::chrono::system_clock::now();
    auto start = now - std::chrono::hours(1);
    auto result = storage_->readSeries(
        series.getLabels(),
        std::chrono::duration_cast<std::chrono::nanoseconds>(start.time_since_epoch()).count(),
        std::chrono::duration_cast<std::chrono::nanoseconds>(now.time_since_epoch()).count()
    );

    ASSERT_TRUE(result);
    EXPECT_EQ(result->getSamples().size(), series.getSamples().size());
}

TEST_F(StorageTest, Recovery) {
    // Write initial data
    auto series = TestUtils::generateTestSeries({
        {"__name__", "test_metric"},
        {"instance", "test-1"}
    }, 100);

    EXPECT_NO_THROW(storage_->writeSeries(series));

    // Close storage
    storage_.reset();

    // Reopen storage
    StorageConfig config{
        .data_path = test_path_.string(),
        .max_block_size = 1024 * 1024
    };
    storage_ = std::make_unique<Storage>(config);

    // Verify data survived
    auto now = std::chrono::system_clock::now();
    auto start = now - std::chrono::hours(1);
    auto result = storage_->readSeries(
        series.getLabels(),
        std::chrono::duration_cast<std::chrono::nanoseconds>(start.time_since_epoch()).count(),
        std::chrono::duration_cast<std::chrono::nanoseconds>(now.time_since_epoch()).count()
    );

    ASSERT_TRUE(result);
    EXPECT_EQ(result->getSamples().size(), series.getSamples().size());
}

} // namespace tsdb::test 